File: app/src/main/java/com/bracketcove/postrainer/data/alarmdatabase/AlarmDatabase.java
package com.bracketcove.postrainer.data.alarmdatabase;

import com.bracketcove.postrainer.data.realmmodel.RealmAlarm;
import com.bracketcove.postrainer.data.viewmodel.Alarm;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Completable;
import io.reactivex.CompletableEmitter;
import io.reactivex.CompletableOnSubscribe;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import io.reactivex.FlowableOnSubscribe;
import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.RealmResults;

/**
 * Created by Ryan on 10/04/2017.
 */

public class AlarmDatabase implements AlarmSource {

    public AlarmDatabase() {

    }

    @Override
    public Completable deleteAlarm(final Alarm alarm) {
        return Completable.create(
                new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(final CompletableEmitter e) throws Exception {
                        Realm realm = Realm.getDefaultInstance();

                        realm.beginTransaction();

                        RealmQuery<RealmAlarm> query = realm.where(RealmAlarm.class);

                        query.equalTo("alarmId", alarm.getAlarmId());
                        RealmResults<RealmAlarm> result = query.findAll();

                        if (result.size() == 0) {
                            realm.cancelTransaction();
                            e.onError(new Exception());
                        } else {
                            result.deleteFromRealm(0);
                            realm.commitTransaction();
                            e.onComplete();
                        }
                    }
                }
        );
    }

    @Override
    public Completable updateAlarm(final Alarm alarm) {
        return Completable.create(
                new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(final CompletableEmitter e) throws Exception {
                        Realm realm = Realm.getDefaultInstance();

                        realm.beginTransaction();

                        RealmAlarm realmAlarm = new RealmAlarm();

                        realmAlarm.setAlarmId(alarm.getAlarmId());
                        realmAlarm.setHourOfDay(alarm.getHourOfDay());
                        realmAlarm.setMinute(alarm.getMinute());
                        realmAlarm.setAlarmTitle(alarm.getAlarmTitle());
                        realmAlarm.setActive(alarm.isActive());
                        realmAlarm.setVibrateOnly(alarm.isVibrateOnly());
                        realmAlarm.setRenewAutomatically(alarm.isRenewAutomatically());

                        realm.copyToRealmOrUpdate(realmAlarm);

                        realm.commitTransaction();

                        e.onComplete();
                    }
                }
        );
    }

    @Override
    public Flowable<List<Alarm>> getAlarms() {
        return Flowable.create(
                new FlowableOnSubscribe<List<Alarm>>() {
                    @Override
                    public void subscribe(FlowableEmitter<List<Alarm>> e) throws Exception {
                        Realm realm = Realm.getDefaultInstance();

                        RealmQuery<RealmAlarm> query = realm.where(RealmAlarm.class);
                        RealmResults<RealmAlarm> result = query.findAll();

                        List<Alarm> alarmList = new ArrayList<>();

                        if (result.size() == 0) {
                            e.onComplete();
                        } else {
                            for (int i = 0; i < result.size(); i++) {
                                Alarm alarm = new Alarm();
                                RealmAlarm realmAlarm = result.get(i);

                                alarm.setActive(realmAlarm.isActive());
                                alarm.setRenewAutomatically(realmAlarm.isRenewAutomatically());
                                alarm.setVibrateOnly(realmAlarm.isVibrateOnly());
                                alarm.setHourOfDay(realmAlarm.getHourOfDay());
                                alarm.setMinute(realmAlarm.getMinute());
                                alarm.setAlarmTitle(realmAlarm.getAlarmTitle());
                                alarm.setAlarmId(realmAlarm.getAlarmId());

                                alarmList.add(
                                        alarm
                                );
                            }
                            e.onNext(alarmList);
                        }
                    }

                },
                BackpressureStrategy.LATEST);
    }

    @Override
    public Flowable<Alarm> getAlarmsById(final String reminderId) {
        return Flowable.create(
                new FlowableOnSubscribe<Alarm>() {
                    @Override
                    public void subscribe(FlowableEmitter<Alarm> e) throws Exception {

                        Realm realm = Realm.getDefaultInstance();

                        RealmQuery<RealmAlarm> query = realm.where(RealmAlarm.class);
                        query.equalTo("alarmId", reminderId);

                        RealmResults<RealmAlarm> result = query.findAll();

                        if (result.size() == 0) {
                            e.onError(new Exception("AlarmNotFoundException"));
                        } else {
                            RealmAlarm realmAlarm = result.get(0);
                            Alarm alarm = new Alarm();

                            alarm.setAlarmId(realmAlarm.getAlarmId());
                            alarm.setActive(realmAlarm.isActive());
                            alarm.setRenewAutomatically(realmAlarm.isRenewAutomatically());
                            alarm.setVibrateOnly(realmAlarm.isVibrateOnly());
                            alarm.setHourOfDay(realmAlarm.getHourOfDay());
                            alarm.setMinute(realmAlarm.getMinute());
                            alarm.setAlarmTitle(realmAlarm.getAlarmTitle());

                            e.onNext(alarm);
                        }
                    }
                },
                BackpressureStrategy.LATEST);
    }
}

File: androiddata/src/main/java/com/wiseassblog/androiddata/data/alarmdatabase/AlarmDatabase.kt
package com.wiseassblog.androiddata.data.alarmdatabase

import com.wiseassblog.androiddata.data.realmmodel.RealmAlarm
import com.wiseassblog.androiddata.data.toAlarm
import com.wiseassblog.androiddata.data.toRealmAlarm
import com.wiseassblog.common.AlarmServiceException
import com.wiseassblog.common.ResultWrapper
import com.wiseassblog.domain.domainmodel.Alarm
import com.wiseassblog.domain.repository.IAlarmRepository
import io.realm.Realm
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Created by Ryan on 10/04/2017.
 */

object AlarmDatabase : IAlarmRepository {
    override suspend fun deleteAlarm(alarm: Alarm): ResultWrapper<Exception, Unit> = withContext(Dispatchers.IO) {
        val realm = Realm.getDefaultInstance()

        //get first alarm which matches ID
        val queryResult = realm.where<RealmAlarm>(RealmAlarm::class.java)
            .equalTo("alarmId", alarm.alarmId)
            .findFirst()

        if (queryResult == null) ResultWrapper.build { throw AlarmServiceException }
        else {
            realm.executeTransaction {
                queryResult.deleteFromRealm()
            }

            ResultWrapper.build { Unit }
        }
    }

    override suspend fun updateAlarm(alarm: Alarm): ResultWrapper<Exception, Unit> = withContext(Dispatchers.IO) {
        val realm = Realm.getDefaultInstance()

        realm.executeTransaction {
            it.insertOrUpdate(alarm.toRealmAlarm)
        }

        ResultWrapper.build { Unit }
    }

    override suspend fun getAlarms(): ResultWrapper<Exception, List<Alarm>> = withContext(Dispatchers.IO) {
        val realm = Realm.getDefaultInstance()

        val queryResult = realm.where<RealmAlarm>(RealmAlarm::class.java).findAll()

        if (queryResult.size == 0) {
            ResultWrapper.build { emptyList<Alarm>() }
        } else {

            val alarmList = mutableListOf<Alarm>()
            queryResult.forEach {
                alarmList.add(it.toAlarm)
            }
            ResultWrapper.build { alarmList }
        }
    }

    override suspend fun getAlarmById(alarmId: String): ResultWrapper<Exception, Alarm> = withContext(Dispatchers.IO) {
        val realm = Realm.getDefaultInstance()

        //get first alarm which matches ID
        val queryResult = realm.where<RealmAlarm>(RealmAlarm::class.java)
            .equalTo("alarmId", alarmId)
            .findFirst()

        if (queryResult == null) ResultWrapper.build { throw AlarmServiceException }
        else ResultWrapper.build { queryResult.toAlarm }

    }
}
